bigtable/client/table_bulk_apply_test.cc

```diff
@@ -0,0 +1,377 @@
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r1.release();
+          }));
+
+  namespace bt = bigtable;
+  bt::Table table(&client, "foo_table");
+
+  EXPECT_THROW(table.BulkApply(bt::BulkMutation(
+                   bt::SingleRowMutation(
+                       "foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+                   bt::SingleRowMutation(
+                       "bar", {bigtable::SetCell("fam", "col", 0, "qux")}))),
+               std::exception);
+}
+
+/// @test Verify that Table::BulkApply() handles a terminated stream.
+TEST(TableBulkApply, CanceledStream) {
+  MockClient client;
+  using namespace ::testing;
+  EXPECT_CALL(client, Stub())
+      .WillRepeatedly(Invoke(
+          [&client]() -> google::bigtable::v2::Bigtable::StubInterface & {
+            return *client.mock_stub;
+          }));
+  namespace btproto = ::google::bigtable::v2;
+
+  // Simulate a stream that returns one success and then terminates.  We expect
+  // the BulkApply() operation to retry the request, because the mutation is in
+  // an undetermined state.  Well, it should retry assuming it is idempotent,
+  // which happens to be the case in this test.
+  auto r1 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r1, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse *r) {
+        {
+          auto &e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r1, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  // Create a second stream returned by the mocks when the client retries.
+  auto r2 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r2, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse *r) {
+        {
+          auto &e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r2, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  EXPECT_CALL(*client.mock_stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r1.release();
+          }))
+      .WillOnce(Invoke(
+          [&r2](grpc::ClientContext *, btproto::MutateRowsRequest const &) {
+            return r2.release();
+          }));
+
+  namespace bt = bigtable;
+  bt::Table table(&client, "foo_table");
+
+  EXPECT_NO_THROW(table.BulkApply(bt::BulkMutation(
+      bt::SingleRowMutation("foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+      bt::SingleRowMutation("bar",
+                            {bigtable::SetCell("fam", "col", 0, "qux")}))));
+}
+
+/// @test Verify that Table::BulkApply() reports correctly on too many errors.
+TEST(TableBulkApply, TooManyFailures) {
+  MockClient client;
```

Comment:
This block of code appears in several tests, as-is.

Why not create a test fixture class, make `MockClient` a member var, and add an `EXPECT_CALL` in the `SetUp()` method so it's always there? This will eliminate some boilerplate.

There are probably other ways you can eliminate code duplication in these tests.

Reply:
Great idea, thanks!